File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D24.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.combinatorics.getPairs
import org.gristle.pdxpuzzles.utilities.parsing.getLongs
import org.gristle.pdxpuzzles.utilities.parsing.getNumbers
import kotlin.math.abs

class Y23D24(private val input: String) : Day {
    
    data class Point3d(val x: Double, val y: Double, val z: Double)

    data class Stone(val pos: Point3d, val vel: Point3d) {
        fun intersects(other: Stone, testArea: ClosedFloatingPointRange<Double>): Boolean {
            // get time for original stone at intersection point
            val t = (other.pos.y + (pos.x - other.pos.x) * other.vel.y / other.vel.x - pos.y) / 
                    (vel.y - vel.x * other.vel.y / other.vel.x)
            if (t < 0.0) return false // intersected in past
            
            // get time for comparison stone at intersection point
            val s = (t * vel.x + pos.x - other.pos.x) / other.vel.x
            if (s < 0.0) return false // intersected in past
            
            // get x and y for intersection point and report whether they are both within the test area 
            val x = t * vel.x + pos.x
            val y = t * vel.y + pos.y
            
            return x in testArea && y in testArea
        }
    }

    private val testArea = 200000000000000.0..400000000000000.0
    
    // H/T Brian Norman for walking through the math. 
    // https://github.com/bnorm/advent-of-code/blob/main/year2023/src/aoc/day24/main.kt#L33-L101
    override fun part1(): Int {
        val stones = input
            .getNumbers { toDoubleOrNull() }
            .chunked(3) { (x, y, z) -> Point3d(x, y, z) }
            .chunked(2) { (pos, vel) -> Stone(pos, vel) }
            .toList()
        return stones.getPairs().count { (a, b) -> a.intersects(b, testArea) } 
    } 

    // This was well beyond my abilities, so I had to look at others' solutions. Roman Elizarov's was the easiest for 
    // me to understand, so I replicate it here in commented, non-golfed fashion.
    // His version relies on the fact that on at least one axis the rock has the same position and velocity as one of
    // the hailstones. This way the bruteforce aspect is kept to a minimum.
    // https://github.com/elizarov/AdventOfCode2023/blob/main/src/Day24_2_golf.kt
    override fun part2(): Long {
        val stones: List<List<List<Long>>> = input.getLongs()
            .chunked(3)
            .chunked(2)
            .toList()
        
        // Get the intersection times for all the hailstones once you find the hailstone that matches the pos/velocity
        // of the rock.
        val times: List<Long> = (0..2).firstNotNullOf { axis ->
            stones.firstNotNullOfOrNull { stone ->
                stones.map { other ->
                    // if the two stones have the same velocity on the tested axis, then further consider whether they
                    // have the same position on the tested axis. If so, they meet immediately at time = 0. 
                    // If not, they are parallel, and it's not a valid position.
                    if (stone[1][axis] == other[1][axis]) {
                        if (stone[0][axis] == other[0][axis]) 0L else return@firstNotNullOfOrNull null
                    } else { // if the stones have a different velocity...
                        // time is an integer (why?), so discard any that don't divide cleanly, which means it will
                        // try the next stone.
                        val diffPos = abs(other[0][axis] - stone[0][axis]) // compute difference in initial positions
                        val diffVel = abs(other[1][axis] - stone[1][axis]) // compute difference in velocities
                        if (diffPos % diffVel == 0L) diffPos / diffVel else return@firstNotNullOfOrNull null
                    }
                }
            }
        } 
        
        // get the indices of the first two hailstones that do not travel parallel  
        val (i, j) = times.asSequence().withIndex()
            .filter { (_, time) -> time > 0 }
            .map { it.index }
            .take(2)
            .toList()
        
        // given a stone, an axis, and a time, calculate the position of the stone on that axis at that time
        fun rockPos(i: Int, axis: Int, time: Long): Long = stones[i][0][axis] + stones[i][1][axis] * time
        
        // for each axis, calculate the position at time 0, and sum it.
        return (0..2).sumOf { axis ->
            val iPos = rockPos(i, axis, times[i])
            iPos - (iPos - rockPos(j, axis, times[j])) / (times[i] - times[j]) * times[i]
        }
    }
}

fun main() = Day.runDay(Y23D24::class)

//    Class creation: 3ms
//    Part 1: 14046 (47ms)
//    Part 2: 808107741406756 (6ms)
//    Total time: 57ms

@Suppress("unused")
private val sampleInput = listOf(
    """19, 13, 30 @ -2,  1, -2
18, 19, 22 @ -1, -1, -2
20, 25, 34 @ -2, -2, -4
12, 31, 28 @ -1, -2, -1
20, 19, 15 @  1, -5, -3
""",
)

File: rust/advent/src/bin/2023/y23d24.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{
    parsing::get_numbers::ContainsNumbers,
    structs::{
        coord::Coord3,
        stopwatch::{ReportDuration, Stopwatch},
    },
};

type Input<'a> = &'a str;

#[derive(Copy, Clone, Debug)]
struct Point {
    x: f64,
    y: f64,
}

#[derive(Copy, Clone, Debug)]
struct Stone {
    pos: Point,
    vel: Point,
}

impl Stone {
    fn intersects(&self, other: &Stone) -> bool {
        let t = (other.pos.y + (self.pos.x - other.pos.x) * other.vel.y / other.vel.x - self.pos.y)
            / (self.vel.y - self.vel.x * other.vel.y / other.vel.x);
        if t < 0.0 {
            return false;
        }

        let s = (t * self.vel.x + self.pos.x - other.pos.x) / other.vel.x;
        if s < 0.0 {
            return false;
        }

        let x = t * self.vel.x + self.pos.x;
        let y = t * self.vel.y + self.pos.y;

        let lower = 200000000000000.0;
        let upper = 400000000000000.0;

        x >= lower && x <= upper && y >= lower && y <= upper
    }
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 24).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn part1(input: Input) -> usize {
    input
        .get_numbers::<i64>()
        .tuples()
        .map(|(x, y, _)| Point {
            x: x as f64,
            y: y as f64,
        })
        .tuples()
        .map(|(pos, vel)| Stone { pos, vel })
        .combinations(2)
        .filter(|combo| {
            let a = combo[0];
            let b = combo[1];
            a.intersects(&b)
        })
        .count()
}

fn part2(input: Input) -> i64 {
    let stones: Vec<(Coord3, Coord3)> = input
        .get_numbers::<i64>()
        .tuples::<(_, _, _)>()
        .map(|(x, y, z)| Coord3::new3d(x, y, z))
        .tuples::<(_, _)>()
        .collect();

    let mut times = Vec::new();

    'outer: for axis in 0..=2 {
        'mid: for &(pos, vel) in stones.iter() {
            for &(other_pos, other_vel) in stones.iter() {
                if vel.get(axis) == other_vel.get(axis) {
                    if pos.get(axis) == other_pos.get(axis) {
                        times.push(0);
                    } else {
                        times.clear();
                        continue 'mid;
                    }
                } else {
                    let diff_pos = (other_pos.get(axis).unwrap() - pos.get(axis).unwrap()).abs();
                    let diff_vel = (other_vel.get(axis).unwrap() - vel.get(axis).unwrap()).abs();
                    if diff_pos % diff_vel == 0 {
                        times.push(diff_pos / diff_vel)
                    } else {
                        times.clear();
                        continue 'mid;
                    }
                }
            }
            if !times.is_empty() {
                break 'outer;
            }
        }
    }

    let (i, j) = times
        .iter()
        .enumerate()
        .filter(|&(_, &time)| time > 0)
        .map(|(idx, _)| idx)
        .take(2)
        .next_tuple()
        .unwrap();

    fn rock_pos(stones: &Vec<(Coord3, Coord3)>, i: usize, axis: usize, time: i64) -> i64 {
        stones[i].0.get(axis).unwrap() + stones[i].1.get(axis).unwrap() * time
    }

    (0..=2)
        .map(|axis| {
            let i_pos = rock_pos(&stones, i, axis, times[i]);
            i_pos
                - (i_pos - rock_pos(&stones, j, axis, times[j])) / (times[i] - times[j]) * times[i]
        })
        .sum()
}

#[test]
fn default() {
    let input = get_input(23, 24).unwrap();
    assert_eq!(14046, part1(&input));
    assert_eq!(808107741406756, part2(&input));
}

// Input parsed (18μs)
// 1. 14046 (1ms)
// 2. 808107741406756 (51μs)
// Total: 1ms
